NDTensors/src/dense/dense.jl

```diff
@@ -20,53 +20,67 @@ struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
   end
 end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+#Start with high information constructors and move to low information constructors
+function Dense{ElT, VecT}() where {ElT, VecT<:AbstractArray{ElT}}
+  return default_storagetype(VecT)(VecT())
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+# Construct from a set of indices
+# This will fail if zero(ElT) is not defined for the ElT
+function Dense{ElT,VecT}(inds::Tuple) where {ElT,VecT<:AbstractArray{ElT}}
+  return default_storagetype(VecT)(zeros(VecT, dim(inds)))
+  #return default_storagetype(VecT)(VecT(undef, dim(inds)))
 end
 
-function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
-  return ElT == ElR ? Dense(data) : Dense(ElR.(data))
+function Dense{ElT, VecT}(::UndefInitializer, inds::Tuple) where {ElT, VecT<:AbstractArray{ElT}}
+  return default_storagetype(VecT)(VecT(undef, dim(inds)))
 end
 
-# Construct from a set of indices
-function Dense{ElT,VecT}(inds) where {ElT,VecT<:AbstractArray{ElT}}
-  return Dense(VecT(dim(inds)))
+function Dense{VecT}() where {VecT<:AbstractArray}
+  default_storagetype(VecT)()
 end
 
-Dense{ElT}(dim::Integer) where {ElT} = Dense(zeros(ElT, dim))
+function Dense{VecT}(inds::Tuple) where {VecT<:AbstractArray}
+  default_storagetype(VecT)(inds)
+end
 
-Dense{ElT}(::UndefInitializer, dim::Integer) where {ElT} = Dense(Vector{ElT}(undef, dim))
+function Dense{VecT}(::UndefInitializer, inds::Tuple) where {VecT<:AbstractArray}
+  default_storagetype(VecT)(undef, inds)
+end
 
-Dense(::Type{ElT}, dim::Integer) where {ElT} = Dense{ElT}(dim)
+# This function is ill-defined. It cannot transform a complex type to real...
+function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
+  Dense{ElR, similartype(typeof(data), ElR)}(data)
+end
 
-Dense(x::ElT, dim::Integer) where {ElT<:Number} = Dense(fill(x, dim))
+Dense{ElT}(dim::Integer) where {ElT <: Number} = default_storagetype(ElT)(zeros(default_datatype(ElT), dim))
 
-Dense(dim::Integer) = Dense(Float64, dim)
+Dense{ElT}() where {ElT} = default_storagetype(ElT)()
 
-Dense(::Type{ElT}, ::UndefInitializer, dim::Integer) where {ElT} = Dense{ElT}(undef, dim)
+Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT} = Dense{ElT, VecT}(data)
 
-Dense(::UndefInitializer, dim::Integer) = Dense(Float64, undef, dim)
+Dense(VecT::Type{<:AbstractArray{ElT}}, dim::Integer) where ElT = Dense{ElT, VecT}((dim,))
 
-Dense{ElT}() where {ElT} = Dense(ElT[])
-Dense(::Type{ElT}) where {ElT} = Dense{ElT}()
+Dense(::Type{ElT}, dim::Integer) where {ElT} = Dense{ElT}(dim)
 
-setdata(D::Dense, ndata) = Dense(ndata)
+Dense(x::ElT, dim::Integer) where {ElT<:Number} = Dense{ElT}((fill(x, dim)))
```

Comment:
This implicitly creates CPU data so for consistency we should think about how to write it in terms of `default_datatype`. Maybe:
```julia
function Dense(x::Number, dim::Integer)
  data = default_datatype(typeof(x))(undef, dim)
  return Dense(fill!(data, x))
end
```